### ISSUE TITLE
FindTypeTagName conventions change if method of creating DocumentStore changes

### ISSUE THREAD

**Issue by johnsimons** [Particular/NServiceBus.RavenDB#177]
## Description

Different document id conventions (specifically the implementation of the `Func<Type, string>` delegate `documentStore.Conventions.FindTypeTagName`) were used depending on whether the DocumentStore was created by NServiceBus, or provided by the end user. This would result in differing document collection names for Saga and Timeout data:

* If the end user provided a DocumentStore, the FindTypeTagName would not be affected, resulting in collection names using Raven's default pluralization scheme, such as `TimeoutDatas` and `TestSagaDatas`.
* If the end user provided a connection string or `ConnectionParameters` object, NServiceBus would instantiate the DocumentStore and modify the FindTypeTagName to return the type's unchanged type name, and in the case of sagas, with the word "Data" removed. This resulted in collection names such as `TimeoutData` and `TestSaga`.

Switching from one method to another could cause data loss by making documents stored with the alternate strategy effectively disappear. Alternately, since Saga UniqueIdentity documents still pointed to the correct document id, a saga may be loaded but then throw an exception when trying to save with the different convention.

## Who's affected

Customers attempting change Raven connection methods between 1) providing a full DocumentStore, and 2) allowing NServiceBus to create the DocumentStore by providing a connection string, providing a ConnectionParameters, or allowing the use of the default `http://localhost:8080/{EndpointName}` connection option.

## Symptoms

* After changing how connection information is provided to NServiceBus, timeouts or saga data appear to disappear.
* After changing how connection information is provided to NServiceBus, saga update exceptions occur related to being unable to change the Id of an existing saga.

## Original Report

**This is a bug that currently exists in v3.1 (this release was pulled)**

If user uses:
```c#
configuration.UsePersistence<RavenDBPersistence>()
                .SetDefaultDocumentStore(documentStore);
```

the `FindTypeTagName` convention should not be overridden.

This was working fine in previous releases.

## Reason for Prioritization

Included in `Project: V6 Launch` because it is bound up with #136, which could cause data loss to a user running IsolatedStorage. But if that was fixed, users specifying their own DocStore to fix the transaction recovery settings could switch from that method to a connection string, and if they did, this bug would come into play, and they would lose any Saga/Timeout data stored with the wrong convention.

**Comment by mikeminutillo** [comment]
There are 2 cases to consider. 

1. I had set my own document store before I updated to this version - then do not override the convention as the previous version did not have these conventions applied. 
2. I did not have my own document store before I updated to this version and now I am switching to use my own document store - then you MUST override the convention because your old documents used the convention.

Actually, go back to an older version. 
- Create an endpoint without overriding the doc store
- Put some data in it
- Update the endpoint to override the doc store
- Can you read the old data?

It seems to me that you wouldn't be able to because the conventions weren't being applied the same way.

**Comment by DavidBoike** [comment]
Could you provide a link to ServiceControl where the custom conventions are being sent? Will include in the test.

**Comment by johnsimons** [comment]
My mistake @DavidBoike, ServiceControl does not use custom conventions, it uses the raven default pluralization convention

**Comment by DavidBoike** [comment]
Gateway is too hard to mock quickly, but I did a test involving sagas, timeouts, subscriptions, and outbox on 3.0.6. Here are the collection names that were generated.

| Provided DocStore    | NSB-created Store  |
|----------------------|--------------------|
| OutboxRecords        | OutboxRecord       |
| SagaUniqueIdentities | SagaUniqueIdentity |
| Subscriptions        | Subscription       |
| TestSagaDatas        | TestSaga           |
| TimeoutDatas         | TimeoutData        |

This is a *very not good* position to be in as it makes it essentially impossible to switch back and forth right now.

Ideally we would have specified no conventions whatsoever and tightly controlled document ids that were generated, but now either version above could exist in any given user's system and we have a backward compatibility problem.

The convention is specified on the DocStore by a `public Func<Type, string>` property. My current idea to solve it is:

* Specify no convention before initializing the store
* Grab a copy of whatever conventions currently exist (raven defaults or defined by user)
* After initialize, replace with a new convention that can delegate to the user/default convention.
* Use the same query the Studio uses to get the names of existing collections (below) and store it in a HashSet.
* Within our convention:
  * If the type is not one of ours, delegate to the stored user/default convention func
  * If the type IS one of ours, evaluate what it would be under both historical conventions (plural and non-plural) and if one of those already exists in the HashSet, return that value. If not (new user or never used that feature) then use whatever the "default" convention should be that we pick going forward.

The collection query is a URL that looks like:

    http://<host>/databases/<databaseName>/terms/Raven/DocumentsByEntityName?field=Tag&pageSize=1024

which on my system, on the test I just ran, returns:

    [
        "OutboxRecord",
        "SagaUniqueIdentity",
        "Subscription",
        "TestSaga",
        "TimeoutData"
    ]

The `Raven/DocumentsByEntityName` index **must** exist on the database and is created internally by Raven, because otherwise the Studio wouldn't work.

**Comment by danielmarbach** [comment]
@Particular/persistence-maintainers It seems that @DavidBoike is working full speed on the https://github.com/Particular/PlatformDevelopment/issues/686 issue and @johnsimons is working on the SC patches this one seems to be stuck?

**Comment by DavidBoike** [comment]
I'm working on this. PR coming today.

**Comment by DavidBoike** [comment]
Fix will be included in 3.0.7
